### PR TITLE
[Merged by Bors] - refactor(CharP): make variables explicit + refactor

### PR DESCRIPTION
**Pull request by grhkm21**
Follows up on Yael's comment in #8612 by making `n` (in `x ^ (p ^ n)`) explicit.

---
- [x] depends on: #8612 Lucas' Theorem

CI should fail in places where the theorems are used, but I will fix them using the CI results.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by grhkm21** [comment]
Only the final commit (36th) is from this PR, the rest is from #8612.

**Comment by YaelDillies** [comment]
Why are you making this depend on #8612 and not the other way around?

**Comment by grhkm21** [comment]
Because that one is tagged maintainer-merge so I thought I shouldn't touch it. Oh also the actual dependency part is I also refactor `add_pow_eq_add_pow_mod_mul_pow_add_pow_div`, if that's what you're asking

**Comment by YaelDillies** [comment]
Oh, feel free to keep on touching PRs after they were maintainer merged. We're going to change the workflow soon to make it clearer that `maintainer-merge` is not a once and for all state.

File: Mathlib/Algebra/CharP/Basic.lean
Comment (YaelDillies) [inline Mathlib/Algebra/CharP/Basic.lean:57]:
```suggestion
variable [CommSemiring R] {p : ℕ} (hp : p.Prime) (x y : R) (n : ℕ)
include hp
```
so that `hp` stays before `x y n`

Comment (grhkm21) [reply Mathlib/Algebra/CharP/Basic.lean]: Oh yes, thank you, `include hp` works well. I also changed the section above.

**Review by YaelDillies — APPROVED** [review]
Thanks! 🚀

maintainer merge

File: Mathlib/Algebra/CharP/Basic.lean
Comment (j-loreaux) [inline Mathlib/Algebra/CharP/Basic.lean:25]: I think even `section`ed or `namespace`d `include`s are advised against, in favor of `include ... in`, but it's possible I'm wrong here. Admittedly the sections are short, so maybe it's fine.

File: Mathlib/Algebra/CharP/Basic.lean
Comment (j-loreaux) [inline Mathlib/Algebra/CharP/Basic.lean:59]: ditto

File: Mathlib/Algebra/CharP/Basic.lean
Comment (j-loreaux) [reply Mathlib/Algebra/CharP/Basic.lean:25]: I checked in the reviewer stream on Zulip, and indeed I think we want to only have `include ... in`, so please either change it to that, or else back to explicitly writing the hypothesis in each statement.

Comment (YaelDillies) [reply Mathlib/Algebra/CharP/Basic.lean:25]: I don't think there is such a consensus, though?

Comment (vihdzp) [reply Mathlib/Algebra/CharP/Basic.lean:25]: The reason using `include` like this is advised against is that it's really easy to accidentally include a hypothesis in a theorem that doesn't need it.

**Comment by vihdzp** [comment]
Are you planning to return to this PR? If not I can take this over.

File: Mathlib/Algebra/CharP/Basic.lean
Comment (YaelDillies) [reply Mathlib/Algebra/CharP/Basic.lean:25]: Yes but all theorems here are about that hypothesis! And there is a linter against unused hypotheses anyway

**Comment by YaelDillies** [comment]
I disagree with the need to remove the `include`.

maintainer merge

**Comment by fpvandoorn** [comment]
We don't have a strong policy against `include` (yet?) and I think an `include` directly after the variable command can be fine in certain cases (if it's not omitted later).

bors merge

**Pull request closed without merging**
